docs/design/02-topologies.md

```diff
@@ -0,0 +1,288 @@
+    type: 'integer' # Optional/default `'string'`. For validation and converting; see below for details.
+    required: true # Optional/default `true`
+
+  another_example: &another_example
+    question: 'Enter anything?'
+
+all:
+  # Both these questions will be asked for `metal configure all`.
+  example_question_identifier: *example_question_identifier
+  another_example: *another_example
+
+group:
+  # Just ask this question for `metal configure group`.
+  another_example: *another_example
+
+node:
+  # Similarly, just ask this for `metal configure node`. Note: the answer will
+  # be saved under `different_name` in the resulting config this time however.
+  different_name: *example_question_identifier
+```
+
+## Notes regarding above format
+
+- By defining all the questions within a `questions` object we can then share
+  questions between different categories without duplication using YAML's
+  ability to refer to other parts of the document.
+
+- Any name key can be used when specifying a question for a particular
+  category. Often the same identifier will be used for the question definition
+  and when specifying a question for a category, e.g. a `cluster_name` question
+  may be defined in `questions` and then used as `cluster_name` under `all`,
+  however this will not necessarily always need to be the case.
+
+## Notes on `type`
+
+The `type` for a question will serve the dual purpose of validating input and
+then converting this before saving.
+
+For the minimal version of this feature possible type values will probably just
+include `string`, `integer`, and `boolean`. These will be necessary as since we
+are accepting user input we won't know what type we should save a given answer
+as, and we will need non-string types when templating.
+
+Later we can add types with more validation here, e.g. we can have an `ip` type
+which will save the value as a string but validate that it is a valid IP first.
+While these more complicated validations could be specified separately by
+adding an optional regular expression field to question configs, I think
+building these in to Metalware would be better since:
+
+- there will only be a limited number of these more advanced types, and I don't
+  think we need the flexibility of allowing more validations to be specified in
+  the repo;
+
+- having these built in to Metalware means we won't need to duplicate
+  validations between repos or places within repos;
+
+- we may later want to do non-regex validation, e.g. a `dns` type could
+  validate that a particular IP entered gives the expected DNS response (though
+  this particular example sounds like overkill).
+
+
+# `metal configure` commands
+
+The `metal configure` commands each take their specified questions and ask
+them, save the results to a particular file within a new
+`/var/lib/metalware/config/` directory, and perform any other actions
+appropriate to the resource being configured.
+
+## `metal configure all`
```

Comment:
Nope, the plan is that only the questions specified for each level in `configure.yaml` will be asked when `configure` is run for that level. So some questions may be asked at multiple levels but only those specified. This is because some things it only makes sense to ask at a particular level, e.g. specifying `cluster_name` at a level other than `all` is only likely to make things confusing and broken. Other things people might want to specify differently at different levels for odd use cases, e.g. a DNS server to use, but asking that question for every node you want to configure is probably overkill for the default topology, so we'll probably leave that at the `all` level by default.